go/nbs/s3_table_persister.go

```diff
@@ -184,3 +198,233 @@ func (s partsByPartNum) Less(i, j int) bool {
+		select {
+		case <-done:
+			return
+		default:
+		}
+
+		etag, err := doUpload()
+		if err != nil {
+			failed <- err
+			return
+		}
+		// Try to send along part info. In the case that the upload was aborted, reading from done allows this worker to exit correctly.
+		select {
+		case sent <- s3UploadedPart{int64(partNum), etag}:
+		case <-done:
+			return
+		}
+	}
+
+	// Concurrently begin sending all parts using sendPart().
+	// First, kick off sending all the copyable parts.
+	partNum := int64(1) // Part numbers are 1-indexed
+	for _, cp := range copies {
+		uploadWg.Add(1)
+		go func(cp copyPart, partNum int64) {
+			sendPart(partNum, func() (etag string, err error) {
+				return s3p.uploadPartCopy(cp.name, cp.srcOffset, cp.srcLen, key, uploadID, partNum)
+			})
+		}(cp, partNum)
+		partNum++
+	}
+
+	// Then, split buff (data from |manuals| and index) into parts and upload those concurrently.
+	numManualParts := getNumParts(uint64(len(buff)), s3p.targetPartSize) // TODO: What if this is too big?
+	for i := uint64(0); i < numManualParts; i++ {
+		start, end := i*s3p.targetPartSize, (i+1)*s3p.targetPartSize
+		if i+1 == numManualParts { // If this is the last part, make sure it includes any overflow
+			end = uint64(len(buff))
+		}
+		uploadWg.Add(1)
+		go func(data []byte, partNum int64) {
+			sendPart(partNum, func() (etag string, err error) {
+				return s3p.uploadPart(data, key, uploadID, partNum)
+			})
+		}(buff[start:end], partNum)
+		partNum++
+	}
+
+	// When all the uploads started above are done, close |sent| and |failed| so that the code below will correctly detect that we're done sending parts and move forward.
+	go func() {
+		uploadWg.Wait()
+		close(sent)
+		close(failed)
+	}()
+
+	// Watch |sent| and |failed| for the results of part uploads. If ever one fails, close |done| to stop all the in-progress or pending sendPart() calls and then bail.
+	multipartUpload := &s3.CompletedMultipartUpload{}
+	var firstFailure error
+	for cont := true; cont; {
+		select {
+		case sentPart, open := <-sent:
+			if open {
+				multipartUpload.Parts = append(multipartUpload.Parts, &s3.CompletedPart{
+					ETag:       aws.String(sentPart.etag),
+					PartNumber: aws.Int64(sentPart.idx),
+				})
+			}
+			cont = open
+
+		case err := <-failed:
+			if err != nil && firstFailure == nil { // nil err may happen when failed gets closed
+				firstFailure = err
+				close(done)
+			}
+		}
+	}
+
+	// If there was any failure detected above, |done| is already closed
+	if firstFailure == nil {
+		close(done)
+	}
+	sort.Sort(partsByPartNum(multipartUpload.Parts)) // S3 requires that these be in part-order
+	return multipartUpload, firstFailure
+}
+
+type copyPart struct {
+	name              string
+	srcOffset, srcLen int64
+}
+
+type manualPart struct {
+	srcR             io.Reader
+	dstStart, dstEnd int64
+}
+
+func dividePlan(plan compactionPlan, minPartSize, maxPartSize uint64) (copies []copyPart, manuals []manualPart, buff []byte) {
```

Comment:
note here that `dividePlan` expects the `plan` sources to ordered by descending data length.